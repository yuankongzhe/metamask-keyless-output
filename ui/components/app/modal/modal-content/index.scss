@use "design-system";

.modal-content {
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 16px 0;

  &__title {
    @include design-system.H3;

    font-weight: 500;
    padding: 16px 0;
    text-align: center;
  }

  &__description {
    @include design-system.H6;

    text-align: center;
  }
}
