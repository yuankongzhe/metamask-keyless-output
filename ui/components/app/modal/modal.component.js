import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Button from '../../ui/button';

/**
 * @deprecated The `<Modal />` component has been deprecated in favor of the new `<Modal>` component from the component-library.
 * Please update your code to use the new `<Modal>` component instead, which can be found at ui/components/component-library/modal/modal.tsx.
 * You can find documentation for the new Modal component in the MetaMask Storybook:
 * {@link https://metamask.github.io/metamask-storybook/?path=/docs/components-componentlibrary-modal--docs}
 * If you would like to help with the replacement of the old Modal component, please submit a pull request
 */

export default class Modal extends PureComponent {
  static propTypes = {
    children: PropTypes.node,
    contentClass: PropTypes.string,
    containerClass: PropTypes.string,
    testId: PropTypes.string,
    // Header text
    headerText: PropTypes.string,
    onClose: PropTypes.func,
    // Submit button (right button)
    onSubmit: PropTypes.func,
    submitType: PropTypes.string,
    submitText: PropTypes.string,
    submitDisabled: PropTypes.bool,
    hideFooter: PropTypes.bool,
    // Cancel button (left button)
    onCancel: PropTypes.func,
    cancelType: PropTypes.string,
    cancelText: PropTypes.string,
  };

  static defaultProps = {
    submitType: 'primary',
    cancelType: 'secondary',
  };

  render() {
    const {
      children,
      headerText,
      onClose,
      onSubmit,
      submitType,
      submitText,
      submitDisabled,
      onCancel,
      cancelType,
      cancelText,
      contentClass,
      containerClass,
      hideFooter,
      testId,
    } = this.props;

    return (
      <div
        className={classnames('modal-container', containerClass)}
        data-testid={testId}
      >
        {headerText && (
          <div className="modal-container__header">
            <div className="modal-container__header-text">{headerText}</div>
            <div
              className="modal-container__header-close"
              data-testid="modal-header-close"
              onClick={onClose}
            />
          </div>
        )}
        <div className={classnames('modal-container__content', contentClass)}>
          {children}
        </div>
        {hideFooter ? null : (
          <div className="modal-container__footer">
            {onCancel && (
              <Button
                type={cancelType}
                onClick={onCancel}
                className="modal-container__footer-button"
              >
                {cancelText}
              </Button>
            )}
            <Button
              type={submitType}
              onClick={onSubmit}
              disabled={submitDisabled}
              className="modal-container__footer-button"
            >
              {submitText}
            </Button>
          </div>
        )}
      </div>
    );
  }
}
