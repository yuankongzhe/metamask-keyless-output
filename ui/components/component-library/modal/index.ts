export { Modal } from './modal';
export type { ModalProps } from './modal.types';
export { useModalContext } from './modal.context';
