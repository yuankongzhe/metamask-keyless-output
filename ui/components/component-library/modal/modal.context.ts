import { createContext, useContext } from 'react';

import type { ModalProps } from './modal.types';

export type ModalContextType = Omit<ModalProps, 'children'>;

export const ModalContext = createContext<ModalContextType | undefined>(
  undefined,
);

export const useModalContext = () => {
  const context = useContext(ModalContext);
  if (!context) {
    throw new Error(
      'useModalContext must be used within a ModalProvider, Seems you forgot to wrap the components in "<Modal />"',
    );
  }
  return context;
};
