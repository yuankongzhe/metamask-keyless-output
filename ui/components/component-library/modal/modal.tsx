import React, { forwardRef, Ref } from 'react';
import ReactDOM from 'react-dom';
import classnames from 'classnames';

import { ModalProps } from './modal.types';
import { ModalContext } from './modal.context';

export const Modal = forwardRef(
  (
    {
      className = '',
      isOpen,
      onClose,
      children,
      isClosedOnOutsideClick = true,
      isClosedOnEscapeKey = true,
      autoFocus = true,
      initialFocusRef,
      finalFocusRef,
      restoreFocus,
      ...props
    }: ModalProps,
    ref: Ref<HTMLDivElement>,
  ) => {
    const context = {
      isOpen,
      onClose,
      isClosedOnOutsideClick,
      isClosedOnEscapeKey,
      autoFocus,
      initialFocusRef,
      finalFocusRef,
      restoreFocus,
    };
    return isOpen
      ? ReactDOM.createPortal(
          <ModalContext.Provider value={context}>
            <div
              className={classnames('mm-modal', className)}
              ref={ref}
              {...props}
            >
              {children}
            </div>
          </ModalContext.Provider>,
          document.body,
        )
      : null;
  },
);

export default Modal;
