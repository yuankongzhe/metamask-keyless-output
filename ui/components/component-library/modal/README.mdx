import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import { Modal } from './modal';

# Modal

The `Modal` focuses the user's attention exclusively on information via a window that is overlaid on primary content. It should be used with the `ModalOverlay`, `ModalContent` and `ModalHeader` components to create a complete modal.

[MetaMask Design System Guides](https://www.notion.so/MetaMask-Design-System-Guides-Design-f86ecc914d6b4eb6873a122b83c12940)

<Canvas>
  <Story id="components-componentlibrary-modal--default-story" />
</Canvas>

## Props

<ArgsTable of={Modal} />

### Usage

The `Modal` component is a very atomic level component that is meant to be used with `ModalOverlay`, `ModalContent` and `ModalHeader`.

When the modal opens:

- Focus is trapped within the modal and set to the first tabbable element.
- Content behind a modal dialog is inert, meaning that users cannot interact with it.
- Use the `isOpen` prop to control whether the modal is open or closed.
- Use the `onClose` prop to fire a callback when the modal is closed. This is used for the `isClosedOnOutsideClick` prop and the `isClosedOnEscapeKey`.

<Canvas>
  <Story id="components-componentlibrary-modal--usage" />
</Canvas>

```jsx
import React, { useState, useRef } from 'react';
import { Modal, ModalOverlay, ModalContent, ModalHeader, ModalBody, ModalFooter, Text, Button } from '../../component-library';

const [open, setOpen] = useState(false);

const handleOnClick = () => {
  setOpen(true);
};

const handleOnClose = () => {
  setOpen(false);
};

<Button onClick={handleOnClick}>OpenModal</Button>
<Modal
  isOpen={open}
  onClose={handleOnClose}
>
  <ModalOverlay />
  <ModalContent>
    <ModalHeader onClose={handleOnClose} onBack={handleOnClose}>
      Modal Header
    </ModalHeader>
    <ModalBody>
      <Text>ModalBody children</Text>
    </ModalBody>
    <ModalFooter onSubmit={handleOnClose} onCancel={handleOnCancel}>
  </ModalContent>
</Modal>
```

### Is Closed On Outside Click

Use the `isClosedOnOutsideClick` prop to control whether the modal should close when the user clicks outside of the modal.

Defaults to `true`.

<Canvas>
  <Story id="components-componentlibrary-modal--is-closed-on-outside-click" />
</Canvas>

```jsx
import { Modal } from '../../component-library';

<Modal isClosedOnOutsideClick={false} />;
```

### Is Closed On Escape Key

Use the `isClosedOnEscapeKey` prop to control whether the modal should close when the user presses the escape key.

Defaults to `true`.

<Canvas>
  <Story id="components-componentlibrary-modal--is-closed-on-escape-key" />
</Canvas>

```jsx
import { Modal } from '../../component-library';

<Modal isClosedOnEscapeKey={false} />;
```

### Initial Focus Ref

Use the `initialFocusRef` to set the `ref` of the element to receive focus initially. This is useful for input elements that should receive focus when the modal opens.

<Canvas>
  <Story id="components-componentlibrary-modal--initial-focus-ref" />
</Canvas>

```jsx
import React, { useState, useRef } from 'react';
import { Modal, ModalOverlay, ModalContent, ModalHeader, ModalBody, TextFieldSearch, Button } from '../../component-library';

// Ref to set initial focus
const inputRef = React.useRef<HTMLDivElement>(null);

const [open, setOpen] = useState(false);

const handleOnClick = () => {
  setOpen(true);
};

const handleOnClose = () => {
  setOpen(false);
};

<Button onClick={handleOnClick}>Open modal</Button>
<Modal
  isOpen={isOpen}
  onClose={handleOnClose}
  initialFocusRef={inputRef}
>
  <ModalOverlay />
  <ModalContent >
    <ModalHeader
      onClose={handleOnClose}
      onBack={handleOnClose}
      marginBottom={4}
    >
      Modal Header
    </ModalHeader>
    <TextFieldSearch
      placeholder="Search"
      inputProps={{ ref: inputRef }}
      width={BLOCK_SIZES.FULL}
    />
    <ModalBody>
      <Text>ModalBody children</Text>
    </ModalBody>
  </ModalContent>
</Modal>
```

### Final Focus Ref

Use the `finalFocusRef` to set the `ref` of the element to receive focus when the modal closes.

<Canvas>
  <Story id="components-componentlibrary-modal--final-focus-ref" />
</Canvas>

```jsx
import React, { useState, useRef } from 'react';
import { Modal, ModalOverlay, ModalContent, ModalHeader, TextFieldSearch, Button } from '../../component-library';

// Ref to set focus after modal closes
const buttonRef = React.useRef<HTMLButtonElement>(null);

const [open, setOpen] = useState(false);

const handleOnClick = () => {
  setOpen(true);
};

const handleOnClose = () => {
  setOpen(false);
};

<Button onClick={handleOnClick} marginRight={4}>
  Open modal
</Button>
<button ref={buttonRef}>Receives focus after close</button>
<Modal
  isOpen={isOpen}
  onClose={handleOnClose}
  finalFocusRef={buttonRef}
>
  <ModalOverlay />
  <ModalContent >
    <ModalHeader
      onClose={handleOnClose}
      onBack={handleOnClose}
      marginBottom={4}
    >
      Modal Header
    </ModalHeader>
    <ModalBody>
      <Text>ModalBody children</Text>
    </ModalBody>
  </ModalContent>
</Modal>
```

### Restore Focus

Use the `restoreFocus` prop to restore focus to the element that triggered the `Modal` once it unmounts

Defaults to `false`

<Canvas>
  <Story id="components-componentlibrary-modal--restore-focus" />
</Canvas>

```jsx
import { Modal } from '../../component-library';

<Modal restoreFocus={true} />;
```

### Auto Focus

If `true`, the first focusable element within the `children` will auto-focused once `Modal` mounts. Depending on the content of `Modal` this is usually the back or close button in the `ModalHeader`.

Defaults to `true`

<Canvas>
  <Story id="components-componentlibrary-modal--auto-focus" />
</Canvas>

```jsx
import { Modal } from '../../component-library';

<Modal autoFocus={false} />;
```

## Accessibility

### Keyboard and Focus Management

- When the modal opens, focus is trapped within it.
- When the modal opens, focus is automatically set to the first enabled element, or the element from `initialFocusRef`.
- When the modal closes, focus returns to the element that was focused before the modal activated, or the element from `finalFocusRef`.
- Clicking on the overlay closes the Modal.
- Pressing ESC closes the Modal.
- Scrolling is blocked on the elements behind the modal.
- The modal is rendered in a portal attached to the end of document.body to break it out of the source order and make it easy to add aria-hidden to its siblings.

### ARIA

- The `ModalContent` has aria-modal="true" and role="dialog"
- The `ModalOverlay` has aria-hidden="true"
