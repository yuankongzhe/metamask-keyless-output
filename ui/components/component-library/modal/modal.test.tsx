/* eslint-disable jest/require-top-level-describe */
import { render, fireEvent } from '@testing-library/react';
import React from 'react';

import { Modal } from './modal';

describe('Modal', () => {
  const onClose = jest.fn();

  beforeEach(() => {
    onClose.mockClear();
  });

  it('should render the Modal without crashing', () => {
    const { getByText, getByTestId } = render(
      <Modal onClose={onClose} isOpen data-testid="modal">
        <div>modal content</div>
      </Modal>,
    );
    expect(getByText('modal content')).toBeDefined();
    expect(getByTestId('modal')).toHaveClass('mm-modal');
  });

  it('should match snapshot', () => {
    const { getByTestId } = render(
      <Modal onClose={onClose} isOpen={true} data-testid="test">
        <div>modal content</div>
      </Modal>,
    );
    expect(getByTestId('test')).toMatchSnapshot();
  });

  it('should render with and additional className', () => {
    const { getByTestId } = render(
      <Modal
        onClose={onClose}
        isOpen
        className="test-class"
        data-testid="modal"
      >
        <div>modal content</div>
      </Modal>,
    );
    expect(getByTestId('modal')).toHaveClass('mm-modal test-class');
  });

  it('should render the modal when isOpen is true', () => {
    const { getByText } = render(
      <Modal isOpen={true} onClose={onClose}>
        <div>modal content</div>
      </Modal>,
    );

    const modalContent = getByText('modal content');
    expect(modalContent).toBeInTheDocument();
  });

  it('should not render the modal when isOpen is false', () => {
    const { queryByText } = render(
      <Modal isOpen={false} onClose={onClose}>
        <div>modal content</div>
      </Modal>,
    );

    const modalContent = queryByText('modal content');
    expect(modalContent).not.toBeInTheDocument();
  });

  it('should call the onClose callback when clicking the close button', () => {
    const { getByText } = render(
      <Modal isOpen={true} onClose={onClose}>
        <div>modal content</div>
        <button onClick={() => onClose()}>Close</button>
      </Modal>,
    );

    const closeButton = getByText('Close');
    fireEvent.click(closeButton);

    expect(onClose).toHaveBeenCalledTimes(1);
  });
});
