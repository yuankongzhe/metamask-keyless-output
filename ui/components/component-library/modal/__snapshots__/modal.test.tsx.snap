// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should match snapshot 1`] = `
<div
  class="mm-modal"
  data-testid="test"
>
  <div>
    modal content
  </div>
</div>
`;
